stdlib/src/collections/deque.mojo

```diff
@@ -0,0 +1,313 @@
+# ===----------------------------------------------------------------------=== #
+# Copyright (c) 2024, Modular Inc. All rights reserved.
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions:
+# https://llvm.org/LICENSE.txt
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ===----------------------------------------------------------------------=== #
+"""Defines the Deque type.
+
+You can import these APIs from the `collections` package.
+
+Examples:
+
+```mojo
+from collections import Deque
+```
+"""
+
+from bit import bit_ceil
+from collections import Optional
+from memory import UnsafePointer
+
+from builtin._documentation import doc_private
+
+# ===----------------------------------------------------------------------===#
+# Deque
+# ===----------------------------------------------------------------------===#
+
+
+struct Deque[ElementType: CollectionElement](
+    Movable, ExplicitlyCopyable, Sized, Boolable
+):
+    """Implements a double-ended queue.
+
+    It supports pushing and popping from both ends in O(1) time resizing the
+    underlying storage as needed.
+
+    Parameters:
+        ElementType: The type of the elements in the deque.
+            Must implement the trait `CollectionElement`.
+    """
+
+    # ===-------------------------------------------------------------------===#
+    # Aliases
+    # ===-------------------------------------------------------------------===#
+
+    alias default_capacity: Int = 64
+    """The default capacity of the deque: must be the power of 2."""
+
+    # ===-------------------------------------------------------------------===#
+    # Fields
+    # ===-------------------------------------------------------------------===#
+
+    var data: UnsafePointer[ElementType]
+    """The underlying storage for the deque."""
+
+    var head: Int
+    """The index of the head: points the first element of the deque."""
+
+    var tail: Int
+    """The index of the tail: points behind the last element of the deque."""
+
+    var capacity: Int
+    """The amount of elements that can fit in the deque without resizing it."""
+
+    var minlen: Int
+    """The minimum required capacity in the number of elements of the deque."""
+
+    var maxlen: Int
+    """The maximum allowed capacity in the number of elements of the deque."""
+
+    var shrink: Bool
+    """The flag defining if the deque storage is re-allocated to make it smaller when possible."""
+
```

Comment:
Agreed, I will make the change. However, if I understand correctly, there’s no inherent safety in Mojo structs (similar to Python classes), meaning "private" `_members` can be modified externally just as easily as "public" `members`. This is somewhat concerning for a language like Mojo, which aspires to be a "systems" language. Other collection implementations vary between using `_data` and `data`, so I opted for the "less cluttered" approach.

Reply:
Yep, you're right that Mojo doesn't have a first-class privacy enforcement capability — all methods and members are directly accessible to all code. The leading underscore is just a convention so that readers of the code can quickly get a sense of which fields are valid to mutate as they see fit, vs fields that should be handled with care.

Not all stdlib types follow that convention consistently, but leading underscores is the direction we're hoping to migrate the standard library incrementally to as we add new types or refactor old types 🙂

Reply:
Bikeshedding: maybe rename to `_auto_shrinking`?

Reply:
Can do this for the private member if prefered, still keeping `shrink` as an argument name in API, yes?

Reply:
What do you think about the public API ctor args names? Now we have this:
- `elements` : the arg taking a List of initial values. Is it fine? Should be a keyword agrument as otherwise it would be taken as a single value of the varargs variant and deque would end up with a List as the only element.
- `capacity` : should be ok, consistent with List implementation
- `min_capacity` : make sense, specific to Deque
- `maxlen` : (as python deque), could be more consistent with `min_capacity` if we rename it to `max_len`, but would diverge from python
- `shrink`: is it good enough? would longer `auto_shrinking` or something else be prefered?